From febe1ccef09f70777b086f938c03f3e71989a7c8 Mon Sep 17 00:00:00 2001
From: Paul Floyd <pjfloyd@wanadoo.fr>
Date: Tue, 7 Jan 2025 08:05:20 +0100
Subject: [PATCH 13/14] Bug 498317 - FdBadUse is not a valid CoreError type in
 a suppression even though it's generated by --gen-suppressions=yes

https://bugs.kde.org/show_bug.cgi?id=498317

(cherry picked from commit 47bdc4a6f3de8e2071561d349fdd5f830388c489)
---
 NEWS                                    | 2 ++
 coregrind/m_errormgr.c                  | 7 +++++--
 coregrind/m_syswrap/syswrap-freebsd.c   | 4 ++++
 none/tests/freebsd/Makefile.am          | 4 +++-
 none/tests/freebsd/bug498317.c          | 7 +++++++
 none/tests/freebsd/bug498317.stderr.exp | 0
 none/tests/freebsd/bug498317.supp       | 8 ++++++++
 none/tests/freebsd/bug498317.vgtest     | 2 ++
 9 files changed, 32 insertions(+), 3 deletions(-)
 create mode 100644 none/tests/freebsd/bug498317.c
 create mode 100644 none/tests/freebsd/bug498317.stderr.exp
 create mode 100644 none/tests/freebsd/bug498317.supp
 create mode 100644 none/tests/freebsd/bug498317.vgtest

diff --git a/NEWS b/NEWS
index a25f9b663098..2fb8ce5c724b 100644
--- a/NEWS
+++ b/NEWS
@@ -8,6 +8,8 @@ The following bugs have been fixed or resolved on this branch.
 489913  WARNING: unhandled amd64-linux syscall: 444 (landlock_create_ruleset)
 494246  syscall fsopen not wrapped
 497130  Recognize new DWARF5 DW_LANG constants
+498317  FdBadUse is not a valid CoreError type in a suppression
+        even though it's generated by --gen-suppressions=yes
 
 To see details of a given bug, visit
   https://bugs.kde.org/show_bug.cgi?id=XXXXXX
diff --git a/coregrind/m_errormgr.c b/coregrind/m_errormgr.c
index 4bbcea02474c..2ce919482f77 100644
--- a/coregrind/m_errormgr.c
+++ b/coregrind/m_errormgr.c
@@ -206,7 +206,8 @@ typedef
       // example should new core errors ever be added.
       ThreadSupp = -1,    /* Matches ThreadErr */
       FdBadCloseSupp = -2,
-      FdNotClosedSupp = -3
+      FdNotClosedSupp = -3,
+      FdBadUseSupp = -4
    }
    CoreSuppKind;
 
@@ -1033,7 +1034,7 @@ static Bool core_error_matches_suppression(const Error* err, const Supp* su)
       return err->ekind == FdBadClose;
    case FdNotClosedSupp:
       return err->ekind == FdNotClosed;
-   case FdBadUse:
+   case FdBadUseSupp:
       return err->ekind == FdBadUse;
    default:
       VG_(umsg)("FATAL: unknown core suppression kind: %d\n", su->skind );
@@ -1522,6 +1523,8 @@ static void load_one_suppressions_file ( Int clo_suppressions_i )
             supp->skind = FdBadCloseSupp;
          else if (VG_STREQ(supp_name, "FdNotClosed"))
             supp->skind = FdNotClosedSupp;
+         else if (VG_STREQ(supp_name, "FdBadUse"))
+            supp->skind = FdBadUseSupp;
          else
             BOMB("unknown core suppression type");
       }
diff --git a/coregrind/m_syswrap/syswrap-freebsd.c b/coregrind/m_syswrap/syswrap-freebsd.c
index 685eb6be076c..a2b79545594e 100644
--- a/coregrind/m_syswrap/syswrap-freebsd.c
+++ b/coregrind/m_syswrap/syswrap-freebsd.c
@@ -1400,6 +1400,10 @@ PRE(sys_fcntl)
       PRINT("sys_fcntl[UNKNOWN] ( %lu, %lu, %lu )", ARG1,ARG2,ARG3);
       I_die_here;
    }
+
+   if (!ML_(fd_allowed)(ARG1, "fcntl", tid, False)) {
+     SET_STATUS_Failure (VKI_EBADF);
+   }
 }
 
 POST(sys_fcntl)
diff --git a/none/tests/freebsd/Makefile.am b/none/tests/freebsd/Makefile.am
index fe4f8db69824..1ccfefb57fe2 100644
--- a/none/tests/freebsd/Makefile.am
+++ b/none/tests/freebsd/Makefile.am
@@ -11,6 +11,8 @@ EXTRA_DIST = \
 	auxv.stderr.exp-freebsd131 \
 	auxv.stderr.exp-freebsd14 \
 	auxv.stderr.exp-arm64 \
+	bug498317.vgtest bug498317.stderr.exp \
+	bug498317.supp \
 	cp.vgtest \
 	cp.stderr.exp \
 	osrel.vgtest \
@@ -61,7 +63,7 @@ EXTRA_DIST = \
 	usrstack.stdout.exp
 
 check_PROGRAMS = \
-	auxv osrel swapcontext hello_world fexecve 452275 usrstack \
+	auxv bug498317 osrel swapcontext hello_world fexecve 452275 usrstack \
 	proc_pid_file sanity_level_thread umtx_shm_creat
 
 AM_CFLAGS   += $(AM_FLAG_M3264_PRI)
diff --git a/none/tests/freebsd/bug498317.c b/none/tests/freebsd/bug498317.c
new file mode 100644
index 000000000000..36a1a5a1365e
--- /dev/null
+++ b/none/tests/freebsd/bug498317.c
@@ -0,0 +1,7 @@
+#include <fcntl.h>
+
+int main(void) {
+        fcntl(-1, F_GETFD);
+        return 0;
+}
+
diff --git a/none/tests/freebsd/bug498317.stderr.exp b/none/tests/freebsd/bug498317.stderr.exp
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/none/tests/freebsd/bug498317.supp b/none/tests/freebsd/bug498317.supp
new file mode 100644
index 000000000000..b3a99447c2a4
--- /dev/null
+++ b/none/tests/freebsd/bug498317.supp
@@ -0,0 +1,8 @@
+{
+   test suppression of FdBadUse
+   CoreError:FdBadUse
+   fun:_fcntl
+   fun:fcntl
+   fun:main
+}
+
diff --git a/none/tests/freebsd/bug498317.vgtest b/none/tests/freebsd/bug498317.vgtest
new file mode 100644
index 000000000000..6579ebce8c56
--- /dev/null
+++ b/none/tests/freebsd/bug498317.vgtest
@@ -0,0 +1,2 @@
+prog: bug498317
+vgopts: -q
-- 
2.47.1

